### PR TITLE
Improve performance of `String::from_str_in`

### PR DESCRIPTION
**Pull request by overlookmotel**
I investigated performance of `String::from_str_in`. Turns out that even after `String::push_str` is optimized (#229), the compiler does not figure out that bounds checks and various calculations can be elided.

So this change makes `from_str_in` about 18% faster for small strings.

PR includes a benchmark to demonstrate this.

Built on top of #229. If you're willing to merge, please merge that one first.

Sorry for sudden barrage of issues and PRs!

### CODE REVIEW

**Comment by overlookmotel** [comment]
CI should pass if rebased on top of #232.

**Review by fitzgen — APPROVED** [review]
Looks good, will merge after the other PR lands. This one probably needs to be rebased as well.

**Comment by overlookmotel** [comment]
Rebased on top of #229 (which is now rebasing on top of main). Have done it that way as otherwise there's merge conflicts. Hope that's OK.

**Comment by fitzgen** [comment]
Thanks!

**Pull request merged**
